### PR TITLE
Add Ubuntu 20.04 to CI and ansible-test

### PR DESCRIPTION
**Pull request by relrod**
##### SUMMARY

What the title says

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

tests

### CODE REVIEW

File: test/integration/targets/collections_runtime_pythonpath/runme.sh [hunk 1]
```diff
@@ -8,6 +8,7 @@ export PIP_DISABLE_PIP_VERSION_CHECK=1
 
 source virtualenv.sh
 
+pip install toml
```
Comment (webknjaz) [inline test/integration/targets/collections_runtime_pythonpath/runme.sh]: Why is this necessary? You shouldn't have to do this here.

Comment (relrod) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh:11]: @webknjaz Actually we need `pip install packaging` too, in order for the test to run by itself, it seems.

Otherwise when I run

```
ansible-test integration collections_runtime_pythonpath --docker ubuntu2004 --python 3.8
```

I get:

```
=== Test that the module gets picked up if installed into site-packages ===
+ python -m pip.__main__ install pep517
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python3/dist-packages/pip/__main__.py", line 16, in <module>
    from pip._internal.cli.main import main as _main  # isort:skip # noqa
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/main.py", line 10, in <module>
    from pip._internal.cli.autocompletion import autocomplete
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/autocompletion.py", line 9, in <module>
    from pip._internal.cli.main_parser import create_main_parser
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/main_parser.py", line 7, in <module>
    from pip._internal.cli import cmdoptions
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/cmdoptions.py", line 26, in <module>
    from pip._internal.models.format_control import FormatControl
  File "/usr/lib/python3/dist-packages/pip/_internal/models/format_control.py", line 4, in <module>
    from pip._vendor.packaging.utils import canonicalize_name
ModuleNotFoundError: No module named 'pip._vendor.packaging'
```

If I explicitly install that, the next error I get is:

```
=== Test that the module gets picked up if installed into site-packages ===
+ python -m pip.__main__ install pep517
Requirement already satisfied: pep517 in /root/ansible/test/results/.tmp/output_dir/venv/lib/python3.8/site-packages (0.8.2)
+ cp -r ansible-collection-python-dist-boo /root/ansible/test/results/.tmp/output_dir/
+ cd /root/ansible/test/results/.tmp/output_dir/ansible-collection-python-dist-boo
+ python -m pep517.build --binary --out-dir dist .
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/root/ansible/test/results/.tmp/output_dir/venv/lib/python3.8/site-packages/pep517/build.py", line 6, in <module>
    import toml
ModuleNotFoundError: No module named 'toml'
```

If I explicitly install both, the test passes.

Comment (relrod) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh:11]: @webknjaz I'm not sure why this is necessary for 20.04 but not 18.04

Comment (relrod) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: This might have ultimately been due to podman betraying me. Let me try reverting this change and see what CI does.

Comment (webknjaz) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: @relrod looks like Ubuntu 20.04 ships broken pip then. `packaging` is vendored in pip upstream.

Comment (webknjaz) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: And `toml` is a direct unconditional dependency of the `pep517` package. So when we `pip install` it below it should pull it in.

Comment (webknjaz) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: > 01:15 + python -m pip.__main__ install pep517
01:15 Requirement already satisfied: pep517 in /root/ansible/test/results/.tmp/output_dir/venv/lib/python3.8/site-packages (0.8.2)

this is weird

Comment (webknjaz) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: `source virtualenv.sh` is supposed to provide a clean venv and it seems like it's not that clean somehow.

Comment (sivel) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: `virtualenv.sh` explicitly uses `--system-site-packages`

Comment (webknjaz) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: Yeah, I see. But I was instructed by @mattclay to use it back when I was working on the PR. So this may mean that ubuntu-20 ships not completely installed dists in system site packages.

Comment (relrod) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: There are several things going on here.

* As @sivel said we use `--system-site-packages`.
* Also in my local tests, `podman` betrayed me, and has an issue where `pip` randomly(!) can't install things: https://gist.githubusercontent.com/relrod/25ebdf7cbdd2b8247349cf180189551f/raw/365cf58441327a6957de76aa45676bba86adf926/gistfile1.txt -- this doesn't reproduce in Docker, just podman. So my previous (now deleted) comments didn't make sense.

I dug into why this is needed a bit more. Ubuntu 18.04's Python3 comes with virtualenv 15.x.x, and 20.04 comes with virtualenv 20.x.x, which has a newer version of setuptools embedded in it which enables pip517 stuff by default. When we enter the virtualenv and ask it to install `pep517`, it bails out because it thinks it has no reason to. But it still doesn't have toml which `python -m pep517.build --binary --out-dir dist .` needs to run successfully. So we do actually need `toml` installed here.

Comment (webknjaz) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: > which has a newer version of setuptools embedded in it which enables pip517 stuff by default. When we enter the virtualenv and ask it to install pep517, it bails out because it thinks it has no reason to. But it still doesn't have toml which `python -m pep517.build --binary --out-dir dist .` needs to run successfully. So we do actually need `toml` installed here.

No, this is not entirely correct. Setuptools does not depend on `pep517` (the package, *not* the standard).

We shouldn't apply band-aid in places where some other bug manifests itself.

Do you have any log entries that show how `pep517` gets into that system site-packages env?

Comment (relrod) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: > Do you have any log entries that show how pep517 gets into that system site-packages env?

Yes. Launch a fresh `ubuntu:20.04` container, install `python3` and `python3-virtualenv`.
Then in the container observe the following:

```
root@322e8c5b896e:/# python3.8 -m virtualenv --system-site-packages --python `which python3.8` venv
created virtual environment CPython3.8.2.final.0-64 in 539ms
  creator CPython3Posix(dest=/venv, clear=False, global=True)
  seeder FromAppData(download=False, pytoml=latest, colorama=latest, distlib=latest, html5lib=latest, contextlib2=latest, setuptools=latest, pkg_resources=latest, retrying=latest, urllib3=latest, msgpack=latest, chardet=latest, lockfile=latest, CacheControl=latest, pep517=latest, idna=latest, progress=latest, pyparsing=latest, six=latest, certifi=latest, requests=latest, distro=latest, wheel=latest, appdirs=latest, webencodings=latest, ipaddr=latest, packaging=latest, pip=latest, via=copy, app_data_dir=/root/.local/share/virtualenv/seed-app-data/v1.0.1.debian)
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator
root@322e8c5b896e:/# source venv/bin/activate
(venv) root@322e8c5b896e:/# pip install pep517
Requirement already satisfied: pep517 in /venv/lib/python3.8/site-packages (0.8.2)
```

This is from a fresh container.

Comment (webknjaz) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: as in `docker run ubuntu:20.04`?

Comment (relrod) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: Yeah. @mattclay and I dug into this a bunch tonight, there's a write-up here: #69203 and this PR is on hold for now until those issues are worked out.

Comment (webknjaz) [reply test/integration/targets/collections_runtime_pythonpath/runme.sh]: I think I'd add `pip check` right before creating the virtualenv in `virtualenv.sh` just to make sure that the promised "clean and ready to use" env is that indeed.

**Comment by SophieDeBenedetto** [comment]
Hi, just checking in on this PR--are there plans to merge soon?

**Comment by geerlingguy** [comment]
As more people are upgrading their systems to Ubuntu 20.04, old builds that used Ansible's official PPA are breaking and more people are switching to Pip installs... from @sivel's [last comment](https://github.com/ansible/ansible/issues/68645#issuecomment-610460321) before closing #68645, he mentioned:

> Packages for new Ubuntu releases are only created during the Ansible release process, and as Focal has yet to be released, a package will not be made available until the next release that is made after Focal has been released.

Focal was released in April, and Ansible has had a number of releases since then... is it possible to get this merged soon? Is it held up on tests?

**Comment by relrod** [comment]
@geerlingguy It's held up by Ubuntu. See https://github.com/ansible/ansible/issues/69203 and https://bugs.launchpad.net/ubuntu/+source/python-virtualenv/+bug/1880749 -- there's not much we can do until that is fixed.

**Comment by zxiiro** [comment]
I see the Ubuntu issue is set to resolved. Are we unblocked here?

**Comment by sivel** [comment]
The issue is fixed in the latest _development_ release of Ubuntu, but not in 20.04.

**Comment by quhsi** [comment]
For those who came here because the PPA does not contain a suitable package,
note that Ansible 2.9.6 is included in [Ubuntu's standard repositories](https://packages.ubuntu.com/focal/ansible).

**Comment by mikerev** [comment]
> For those who came here because the PPA does not contain a suitable package,
> note that Ansible 2.9.6 is included in [Ubuntu's standard repositories](https://packages.ubuntu.com/focal/ansible).

Broken link.

**Comment by mikerev** [comment]
> > > For those who came here because the PPA does not contain a suitable package,
> > > note that Ansible 2.9.6 is included in [Ubuntu's standard repositories](https://packages.ubuntu.com/focal/ansible).
> > 
> > 
> > Broken link.
> 
> I cannot confirm that. Broken network?
> 
> Screenshot

Yeah sorry the server was picking up but requests returned something in the 500's, link was fine. All good now tho.

**Comment by DWSR** [comment]
Bump on this. Ideally I'd love to avoid using something like `pipx` to install Ansible on Focal. What still needs to be done here?

File: test/integration/targets/apt/tasks/main.yml [hunk 2]
```diff
@@ -23,6 +23,7 @@
   - import_tasks: 'apt-multiarch.yml'
     when:
       - ansible_userspace_architecture != apt_foreign_arch
+      - not (ansible_distribution == 'Ubuntu' and ansible_distribution_version == '20.04')
```
Comment (mattclay) [inline test/integration/targets/apt/tasks/main.yml]: Can you include a comment explaining why this is being done?

Comment (relrod) [reply test/integration/targets/apt/tasks/main.yml]: I'm not entirely sure this line is even necessary yet, I just wanted to see what else failed.

I'm going to try replacing the tested package with a different one and see if it works.
